src/rust/rust-proto-new/src/graplinc/common/v1beta1.rs

```diff
@@ -0,0 +1,230 @@
+pub use std::time::{
+    Duration,
+    SystemTime,
+};
+use std::time::{
+    SystemTimeError,
+    UNIX_EPOCH,
+};
+
+use bytes::{
+    Buf,
+    BufMut,
+};
+use prost::Message;
+pub use uuid::Uuid;
+
+use crate::{
+    protobufs::graplinc::common::v1beta1::{
+        Duration as _Duration,
+        Timestamp as _Timestamp,
+        Uuid as _Uuid,
+    },
+    type_url,
+    SerDe,
+    SerDeError,
+};
+
+//
+// Uuid
+//
+
+impl From<_Uuid> for Uuid {
+    fn from(uuid_proto: _Uuid) -> Self {
+        Uuid::from_u128_le(u128::from(uuid_proto.lsb) + (u128::from(uuid_proto.msb) << 64))
+    }
+}
+
+impl From<Uuid> for _Uuid {
+    fn from(uuid: Uuid) -> Self {
+        let bytes = uuid.as_bytes();
+
+        let mut lower: [u8; 8] = Default::default();
+        lower.copy_from_slice(&bytes[..8]);
+
+        let mut upper: [u8; 8] = Default::default();
+        upper.copy_from_slice(&bytes[8..]);
+
+        _Uuid {
+            lsb: u64::from_le_bytes(lower),
+            msb: u64::from_le_bytes(upper),
+        }
+    }
+}
+
+impl type_url::TypeUrl for Uuid {
+    const TYPE_URL: &'static str = "graplsecurity.com/graplinc.common.v1beta1.Uuid";
+}
+
+impl SerDe for Uuid {
+    fn serialize<B>(self, buf: &mut B) -> Result<(), SerDeError>
+    where
+        B: BufMut,
+    {
+        _Uuid::from(self).encode(buf)?;
+        Ok(())
+    }
+
+    fn deserialize<B>(buf: B) -> Result<Self, SerDeError>
+    where
+        B: Buf,
+        Self: Sized,
+    {
+        let uuid_proto: _Uuid = Message::decode(buf)?;
+        Ok(uuid_proto.into())
+    }
+}
+
+//
+// Duration
+//
+
+impl From<_Duration> for Duration {
+    fn from(duration_proto: _Duration) -> Self {
+        Duration::new(duration_proto.seconds, duration_proto.nanos)
+    }
+}
+
+impl From<Duration> for _Duration {
+    fn from(duration: Duration) -> Self {
+        _Duration {
+            seconds: duration.as_secs(),
+            nanos: duration.subsec_nanos(),
+        }
+    }
+}
+
+impl type_url::TypeUrl for Duration {
+    const TYPE_URL: &'static str = "graplsecurity.com/graplinc.common.v1beta1.Duration";
+}
+
+impl SerDe for Duration {
+    fn serialize<B>(self, buf: &mut B) -> Result<(), SerDeError>
+    where
+        B: BufMut,
+    {
+        _Duration::from(self).encode(buf)?;
+        Ok(())
+    }
+
+    fn deserialize<B>(buf: B) -> Result<Self, SerDeError>
+    where
+        B: Buf,
+        Self: Sized,
+    {
+        let duration_proto: _Duration = Message::decode(buf)?;
+        Ok(duration_proto.into())
+    }
+}
+
+//
+// SystemTime (a.k.a. Timestamp)
+//
+
+impl TryFrom<_Timestamp> for SystemTime {
+    type Error = SerDeError;
+
+    fn try_from(timestamp_proto: _Timestamp) -> Result<Self, Self::Error> {
+        match timestamp_proto.duration {
+            Some(
+                crate::protobufs::graplinc::common::v1beta1::timestamp::Duration::BeforeEpoch(
```

Comment:
can we use _Duration::BeforeEpoch? Same for Since

Reply:
for some reason we can't do this, because we get the following error (which admittedly I don't understand): ```
error[E0599]: no associated item named `BeforeEpoch` found for struct `protobufs::graplinc::common::v1beta1::Duration` in the current scope
   --> rust-proto-new/src/graplinc/common/v1beta1.rs:167:32
    |
167 |                     _Duration::BeforeEpoch(
    |                                ^^^^^^^^^^^ associated item not found in `protobufs::graplinc::common::v1beta1::Duration`
    |
   ::: /home/jgrillo/src/grapl/src/rust/target/debug/build/rust-proto-new-f6e7cf9c0a6e6e98/out/graplinc.common.v1beta1.rs:28:1
    |
28  | pub struct Duration {
    | ------------------- associated item `BeforeEpoch` not found for this
```
I also tried to import `BeforeEpoch` and `SinceEpoch` directly e.g. ```
use crate::protobufs::graplinc::common::v1beta1::Duration::{
    BeforeEpoch as _BeforeEpoch,
    SinceEpoch as _SinceEpoch
};```
but that yielded the following error:
```
error[E0432]: unresolved imports `crate::protobufs::graplinc::common::v1beta1::Duration`, `crate::protobufs::graplinc::common::v1beta1::Duration`
  --> rust-proto-new/src/graplinc/common/v1beta1.rs:20:9
   |
20 |         Duration::BeforeEpoch as _BeforeEpoch,
   |         ^^^^^^^^ `Duration` is a struct, not a module
21 |         Duration::SinceEpoch as _SinceEpoch,
   |         ^^^^^^^^ `Duration` is a struct, not a module```
so.. I agree fully qualifying here is _ugly_ but I don't know what else to do.